fetch.bs

```diff
@@ -3225,7 +3238,112 @@ through TLS using ALPN. The protocol cannot be spoofed through HTTP requests in
 </ol>
 </div>
 
+<h3 id=https-upgrades>HTTPS upgrading</h3>
+
+<div algorithm>
+
+<p>User agents may optionally upgrade requests with <a>potentially untrustworthy URL</a>s
+to attempt to fetch them over <a>potentially trustworthy URL</a>s. If an upgraded request
+fails with a network error, it is retried over the original URL.
+
+<p>The HTTPS upgrading algorithm consists of Upgrade and Fallback algorithms, with the following
+steps:
+
+<h4 id=https-upgrades-upgrade>Upgrade algorithm</h4>
+
+<p>Given a request <var>request</var>, this algorithm will rewrite its URL if the request should be
+modified and loaded over <code>"https"</code>, via the following algorithm:
+
+<ol>
+  <li>
+    <p>If one or more of the following conditions are met, return without modifying request:
+    <ul>
+      <li><p><var>request</var> is not a navigation request whose destination is document</li>
+
+      <li><p><var>request</var>'s method is not `<code>GET</code>`</li>
+
+      <li><p><var>request</var>'s URL's scheme is not <code>"http"</code></li>
+
+      <li><p><var>request</var>'s URL's host is non-unique (i.e., contains a non-registrable or
+      non-assignable domain name such as .local or an IP address that falls in a range reserved for
+      non-publicly routable networks)</li>
```

Comment:
"Uniqueness" of a URL isn't a concept that really exists yet. I think it would be best to introduce that concept in URL rather than Fetch, and then reference it from here. I also think you'll want to be a little more exhaustive with the definition. :) You can partially lean on @letitz's work with Private Network Access for some of the IP addresses, but you'll need to specify the TLD matching more robustly (e.g. `.localhost`? `.onion`? other reserved names from https://datatracker.ietf.org/doc/html/rfc6762#appendix-G?)

Reply:
This is another case where it might be better to just have a per-UA carveout (implementation-defined) for exempting certain hostnames from upgrading (Line 3271 below) which would subsume this.

Reply:
Done (removed the separate line item and reprhased this as an implementation-defined exclusion.